fdbrpc/sim2.actor.cpp

```diff
@@ -1727,6 +1739,202 @@ class Sim2 final : public ISimulator, public INetworkConnections {
+				return res;
+			}
+			peerProcess = iter->second;
+		}
+		if (!peerSocket.present() || peerSocket.get()->isClosed()) {
+			peerSocket.reset();
+			auto iter = peerProcess.get()->boundUDPSockets.find(peerAddress.get());
+			if (iter == peerProcess.get()->boundUDPSockets.end()) {
+				return fmap([sz](Void){ return sz; }, delay(0.0));
+			}
+			peerSocket = iter->second.castTo<UDPSimSocket>();
+			// the notation of leaking connections doesn't make much sense in the context of UDP
+			// so we simply handle those in the simulator
+			actors.add(cleanupPeerSocket(this));
+		}
+		if (randomDropPacket()) {
+			return res;
+		}
+		actors.add(send(this, peerSocket.get(), begin, end));
+		return res;
+	}
+	Future<int> sendTo(uint8_t const* begin, uint8_t const* end, NetworkAddress const& peer) override {
+		int sz = int(end - begin);
+		auto res = fmap([sz](Void){ return sz; }, delay(0.0));
+		ASSERT(sz <= MAX_PACKET_SIZE);
+		ISimulator::ProcessInfo* peerProcess = nullptr;
+		Reference<UDPSimSocket> peerSocket;
+		{
+			auto iter = g_sim2.addressMap.find(peer);
+			if (iter == g_sim2.addressMap.end()) {
+				return res;
+			}
+			peerProcess = iter->second;
+		}
+		{
+			auto iter = peerProcess->boundUDPSockets.find(peer);
+			if (iter == peerProcess->boundUDPSockets.end()) {
+				return res;
+			}
+			peerSocket = iter->second.castTo<UDPSimSocket>();
+		}
+		actors.add(send(this, peerSocket, begin, end));
+		return res;
+	}
+	Future<int> receive(uint8_t* begin, uint8_t* end) override {
+		return receiveFrom(begin, end, nullptr);
+	}
+	Future<int> receiveFrom(uint8_t* begin, uint8_t* end, NetworkAddress* sender) override {
+		if (!recvBuffer.empty()) {
+			auto buf = recvBuffer.front().second;
+			if (sender) {
+				*sender = recvBuffer.front().first;
+			}
+			int sz = buf->size();
+			ASSERT(sz <= end - begin);
+			std::copy(buf->begin(), buf->end(), begin);
+			auto res = fmap([sz](Void){ return sz; }, delay(0.0));
+			recvBuffer.pop_front();
+			return res;
+		}
+		return receiveFrom(this, begin, end, sender);
+	}
+	void bind(NetworkAddress const& addr) override {
+		g_sim2.addressMap.erase(_localAddress);
+		process->boundUDPSockets.erase(_localAddress);
+		process->boundUDPSockets.emplace(addr, Reference<UDPSimSocket>::addRef(this));
+		_localAddress = addr;
+		g_sim2.addressMap.emplace(_localAddress, process);
+	}
+
+	NetworkAddress localAddress() const override {
+		return _localAddress;
+	}
+
+};
+
+Future<Reference<IUDPSocket>> Sim2::createUDPSocket(NetworkAddress toAddr) {
+		NetworkAddress localAddress;
+		auto process = g_simulator.getCurrentProcess();
+		if (process->address.ip.isV6()) {
+			IPAddress::IPAddressStore store = process->address.ip.toV6();
+			uint16_t* ipParts = (uint16_t*)store.data();
+			ipParts[7] += deterministicRandom()->randomInt(0, 256);
+			localAddress.ip = IPAddress(store);
+		} else {
+			localAddress.ip = IPAddress(process->address.ip.toV4() + deterministicRandom()->randomInt(0, 256));
```

Comment:
I'm a little confused about the intent here -- is it just to pick a random IP address for the simulated "source machine" to be able to consistently identify it?

Reply:
we don't bind the socket here. So initially it should be bound to a random address and port (this is the same logic that you can find in `connect` for TCP connections).

The reason we want a different IP is to make sure that a user doesn't wrongly identify machines with IP addresses. For our use-cases this is generally unimportant, but in general, this could catch some serious bugs.

Reply:
Should it be `randomInt(1, 256)` then?

Reply:
we still want to have these ports to be on the same port sometimes - otherwise someone might assume that no two connections can share the same address... Basically we just want all possibilities to be tested with some probability